src/app/services/downloader.service.ts

```diff
@@ -0,0 +1,113 @@
+      this.abortSignal = this.abortController.signal;
+      this.totalSize = downloadObject.size;
+
+      // Define the endcoding type.
+      if(downloadObject.sourceEncoding != undefined && downloadObject.targetEncoding != undefined){
+        let queriesObject =
+          {
+           "source": downloadObject.sourceEncoding,
+           "target": downloadObject.targetEncoding
+          };
+        
+        fetchPath = fetchPath+"?"+await this.getQueryString(queriesObject);
+      }
+
+      this.startTime = new Date().getTime();
+      const response = await fetch(fetchPath, {signal: this.abortSignal})
+
+      // Mock size for now
+      // this.totalSize =  Number(response.headers.get('X-zowe-filesize'));
+
+      // TODO: The following core download logic is from the FTA & may require refactoring or future bug-proofing
+      // get the stream from the resposnse body.
+      const readbleStream = response.body != null ? response.body : Promise.reject("Cannot receive data from the host machine");
+      // queueing strategy.
+      const queuingStrategy = new CountQueuingStrategy({ highWaterMark: 5 });
+      // for browsers not supporting writablestram make sure to assign the polyfil writablestream.
+      streamSaver.WritableStream = WritableStream;
+      // create the write stream.
+      const fileStream = streamSaver.createWriteStream(fileName, {
+        writableStrategy: queuingStrategy,
+        readableStrategy: queuingStrategy
+      });
+      const writer = fileStream.getWriter();
+      this.currentWriter = writer;
+      const context = this;
+
+      await new Promise(async resolve => {
+        new ReadableStream({
+          start(controller) {
+            const reader = response.body.getReader();
+            read();
+            function read() {
+              reader.read().then(({done, value}) => {
+                if (done) { // If download completes...
+                  writer.close();
+                  controller.close();
+                  context.log.debug("Finished writing the content to the target file " + fileName + " in host machine. Cleaning up...");
+                  resolve();
+                }
+                if(value != undefined){
+                  writer.write(value);
+                  read();
+                }
+              }).catch(error => {
+                context.log.severe("An error occurred downloading " + fileName + " : ", error)
+                controller.error(error);   
+                resolve(error);               
+              })
+            }
+          }
+        },queuingStrategy);
+      });
+    }
+
+    // Create query strings to append in the request.
+    getQueryString(queries){
+      return Object.keys(queries).reduce((result, key) => {
+          return [...result, `${encodeURIComponent(key)}=${encodeURIComponent(ConfigVariables[queries[key]])}`]
+      }, []).join('&');
+    };
+
+    // Cancel current download.
+    cancelDownload(): void {
```

Comment:
Is `cancelDownload` used anywhere?

Reply:
No, it's not. But it can definitely be used for when we add a queue feature in the future (see "TODO")

Reply:
It will work functionally but has no UI yet until we make it